models/ctsm/R/write_params-clm.R

```diff
@@ -1,670 +1,1284 @@
-      # double fates_allom_latosa_int(fates_pft) ;
-      # fates_allom_latosa_int:long_name = "Leaf area to sap area ratio, intercept [m2/cm2]" ;
-      #fates_allom_latosa_int:units = "ratio" ;
-      # double fates_allom_latosa_slp(fates_pft) ;
-      # fates_allom_latosa_slp:long_name = "Leaf area to sap area ratio, slope (optional)" ;
-      # fates_allom_latosa_slp:units = "unitless" ;
-      # fates_allom_latosa_int = 0.001, 0.001 ;
-      # fates_allom_latosa_slp = 0, 0 ;
-      if(var == "sapwood_ratio"){         
-        ncdf4::ncvar_put(nc=fates.param.nc, varid='fates_allom_latosa_int', start = ipft, count = 1,
-                         vals=udunits2::ud.convert(pft[v],"m2 m-2","m2 cm-2"))
-      }
-      if(var == "leaf_width"){            # Characteristic leaf dimension use for aerodynamic resistance
-        ncdf4::ncvar_put(nc=fates.param.nc, varid='fates_dleaf', start = ipft, count = 1,
-                         vals=udunits2::ud.convert(pft[v],"mm","m"))
-        #PEcAn.logger::logger.debug(paste0("fates_dleaf: ",udunits2::ud.convert(pft[v],"mm","m"))) # temp debugging
-      }
-      ## Currently not in param.nc file despite being on NGEE-T parameter list       
-      #       if(var == "nonlocal_dispersal"){    # Place-holder parameter for important seed dispersal parameters
-      #         ncvar_put(nc=param.nc, varid='seed_dispersal_x', start = ipft, count = 1,
-      #                   vals=pft[v])
-      #       }
-      if(var == "hgt_min"){               # the minimum height (ie starting height) of a newly recruited plant" ;
-        ncdf4::ncvar_put(nc=fates.param.nc, varid='fates_hgt_min', start = ipft, count = 1,
-                         vals=pft[v])
-      }
-      if(var == "leaf_reflect_nir"){      # Leaf reflectance: near-IR	[0-1]
-        ncdf4::ncvar_put(nc=fates.param.nc, varid='fates_rholnir', start = ipft, count = 1,
-                         vals=pft[v])
-      }
-      if(var == "leaf_reflect_vis"){      # Leaf reflectance: visible	[0-1]
-        ncdf4::ncvar_put(nc=fates.param.nc, varid='fates_rholvis', start = ipft, count = 1,
-                         vals=pft[v])
-      }
-      if(var == "wood_reflect_nir"){      # Stem reflectance: near-IR	[0-1]
-        ncdf4::ncvar_put(nc=fates.param.nc, varid='fates_rhosnir', start = ipft, count = 1,
-                         vals=pft[v])
-      }
+    ## Loop over PFTS
+    # for testing
+    # trait.values <- list(c4_grass = list(vmax = 1, b0 = 2))
+    npft <- length(trait.values)
+    PEcAn.logger::logger.debug('there are ',
+                               npft,
+                               'PFTs in this run, they are named:',
+                               names(trait.values))
+    ctsm_pftnames <-
+      stringr::str_trim(tolower(ncdf4::ncvar_get(ctsm.param.nc, "pftname")))
+    for (i in seq_len(npft)) {
+      pft <- trait.values[[i]]
+      pft.name <- names(trait.values)[i]
+      if (is.null(pft.name) | is.na(pft.name)) {
```

Comment:
Use double `&&` and `||` in if-else statements.

Also, not strictly necessary, but w.r.t. `[i]` on the previous line, I've gotten into the habit of using `[[` for all subsetting where I'm expecting to return only a single value, even for atomic vectors. A nice advantage of that is that accidentally passing a vector (length > 1) to `[[` will throw an error, so it's a nice failsafe for extracting values that you know have to be scalars. It's definitely more a quirk of my own coding style rather than a convention we should adapt, but just thought I'd throw it out there.